File: js/normalisation.js
/*
 * Normalisation des 21 landmarks de la main → vecteur de 63 nombres.
 *
 * C'est la clé de la robustesse : après normalisation, le vecteur ne dépend
 * plus de la position de la main dans l'image, de sa distance à la caméra,
 * ni de l'inclinaison du poignet. Trois opérations :
 *
 *   1. TRANSLATION : on place l'origine sur le POIGNET (point 0) ;
 *   2. ÉCHELLE     : on divise par la distance poignet → base du majeur
 *                    (point 9), une longueur stable de la main ;
 *   3. ROTATION    : on exprime les points dans un repère attaché à la main
 *                    (axe Y = poignet→majeur, axe Z = normale de la paume).
 *
 * En plus : les mains GAUCHES sont miroitées pour ressembler aux droites,
 * afin qu'un même signe donne le même vecteur quelle que soit la main.
 */

const POIGNET = 0
const BASE_MAJEUR = 9
const BASE_INDEX = 5
const BASE_AURICULAIRE = 17

/*
 * Empreinte de la convention de normalisation. Le modèle entraîné en Python
 * porte la même empreinte ; si elles diffèrent, c'est que l'entraînement et
 * l'inférence ne parlent pas le même langage → on refuse le modèle plutôt
 * que de prédire n'importe quoi. À incrémenter si on change la normalisation.
 */
export const VERSION_NORMALISATION = 'main-63d-v2'

// Petites fonctions de calcul vectoriel 3D (lisibles, sans bibliothèque).
function soustraire(a, b) {
  return [a.x - b.x, a.y - b.y, a.z - b.z]
}
function produitScalaire(a, b) {
  return a[0] * b[0] + a[1] * b[1] + a[2] * b[2]
}
function produitVectoriel(a, b) {
  return [
    a[1] * b[2] - a[2] * b[1],
    a[2] * b[0] - a[0] * b[2],
    a[0] * b[1] - a[1] * b[0],
  ]
}
function longueur(a) {
  return Math.hypot(a[0], a[1], a[2])
}
function normaliserVecteur(a) {
  const n = longueur(a)
  return n > 1e-9 ? [a[0] / n, a[1] / n, a[2] / n] : [0, 0, 0]
}

/**
 * @param {{x:number,y:number,z:number}[]} landmarks — 21 points MediaPipe.
 * @param {'Left'|'Right'} lateralite — main détectée (étiquette MediaPipe).
 * @returns {Float32Array|null} vecteur 63-dim, ou null si géométrie invalide.
 */
export function normaliserMain(landmarks, lateralite = 'Right') {
  if (!landmarks || landmarks.length < 21) return null

  const poignet = landmarks[POIGNET]
  const majeur = landmarks[BASE_MAJEUR]

  // Échelle de référence : distance poignet → base du majeur.
  const versMajeur = soustraire(majeur, poignet)
  const echelle = longueur(versMajeur)
  if (echelle < 1e-6) return null // main dégénérée (mauvaise détection)

  // Repère canonique de la main.
  const axeY = normaliserVecteur(versMajeur)
  const traversPaume = soustraire(landmarks[BASE_INDEX], landmarks[BASE_AURICULAIRE])
  const axeZ = normaliserVecteur(produitVectoriel(traversPaume, axeY)) // normale paume
  const axeX = normaliserVecteur(produitVectoriel(axeY, axeZ))
  if (longueur(axeZ) < 1e-6) return null

  // Miroir gauche → droite : on inverse l'axe X.
  const miroir = lateralite === 'Left' ? -1 : 1

  const vecteur = new Float32Array(63)
  for (let i = 0; i < 21; i++) {
    const relatif = soustraire(landmarks[i], poignet)
    vecteur[i * 3] = (produitScalaire(relatif, axeX) / echelle) 
    vecteur[i * 3 + 1] = produitScalaire(relatif, axeY) / echelle
    vecteur[i * 3 + 2] = produitScalaire(relatif, axeZ) / echelle* miroir
  }
  return vecteur
}

File: entrainement/normalisation.py
"""
Normalisation des 21 landmarks de la main → vecteur de 63 nombres.

PORT EXACT de js/normalisation.js. Les deux DOIVENT produire le même vecteur
pour les mêmes landmarks, sinon le modèle entraîné en Python ne veut rien dire
pour l'inférence dans le navigateur (voir entrainement/test_coherence.py).

Trois opérations, identiques au JS :
  1. TRANSLATION : origine sur le POIGNET (point 0) ;
  2. ÉCHELLE     : division par la distance poignet → base du majeur (point 9) ;
  3. ROTATION    : projection dans un repère attaché à la main
   a                (axe Y = poignet→majeur, axe Z = normale de la paume).
En plus : les mains GAUCHES sont miroitées sur les droites (axe X inversé).

On reste en Python pur (pas de numpy) pour effectuer EXACTEMENT les mêmes
opérations scalaires que le JS. Le vecteur est ensuite arrondi en float32,
comme le Float32Array du navigateur.
"""

import math
import struct

VERSION_NORMALISATION = "main-63d-v2"

POIGNET = 0
BASE_MAJEUR = 9
BASE_INDEX = 5
BASE_AURICULAIRE = 17


def _soustraire(a, b):
    return [a[0] - b[0], a[1] - b[1], a[2] - b[2]]


def _produit_scalaire(a, b):
    return a[0] * b[0] + a[1] * b[1] + a[2] * b[2]


def _produit_vectoriel(a, b):
    return [
        a[1] * b[2] - a[2] * b[1],
        a[2] * b[0] - a[0] * b[2],
        a[0] * b[1] - a[1] * b[0],
    ]


def _longueur(a):
    return math.hypot(a[0], a[1], a[2])


def _normaliser_vecteur(a):
    n = _longueur(a)
    if n > 1e-9:
        return [a[0] / n, a[1] / n, a[2] / n]
    return [0.0, 0.0, 0.0]


def _en_float32(x):
    """Arrondit un float64 au float32 le plus proche, comme un Float32Array JS."""
    return struct.unpack("f", struct.pack("f", x))[0]


def normaliser_main(landmarks, lateralite="Right"):
    """
    landmarks : liste de 21 points, chacun (x, y, z) — tuple/liste ou objet
                avec .x/.y/.z. lateralite : 'Left' ou 'Right'.
    Retourne une liste de 63 floats (float32), ou None si géométrie invalide.
    """
    if landmarks is None or len(landmarks) < 21:
        return None

    pts = [_en_xyz(p) for p in landmarks]
    poignet = pts[POIGNET]
    majeur = pts[BASE_MAJEUR]

    vers_majeur = _soustraire(majeur, poignet)
    echelle = _longueur(vers_majeur)
    if echelle < 1e-6:
        return None

    axe_y = _normaliser_vecteur(vers_majeur)
    travers_paume = _soustraire(pts[BASE_INDEX], pts[BASE_AURICULAIRE])
    axe_z = _normaliser_vecteur(_produit_vectoriel(travers_paume, axe_y))
    axe_x = _normaliser_vecteur(_produit_vectoriel(axe_y, axe_z))
    if _longueur(axe_z) < 1e-6:
        return None

    miroir = -1.0 if lateralite == "Left" else 1.0

    vecteur = [0.0] * 63
    for i in range(21):
        relatif = _soustraire(pts[i], poignet)
        vecteur[i * 3]     = _en_float32(_produit_scalaire(relatif, axe_x) / echelle)
        vecteur[i * 3 + 1] = _en_float32(_produit_scalaire(relatif, axe_y) / echelle)
        vecteur[i * 3 + 2] = _en_float32((_produit_scalaire(relatif, axe_z) / echelle) * miroir)
    return vecteur


def _en_xyz(p):
    """Accepte un point sous forme (x,y,z), [x,y,z] ou objet .x/.y/.z."""
    if hasattr(p, "x"):
        return (p.x, p.y, p.z)
    return (p[0], p[1], p[2])
